common/src/app/common/types/container.cljc

```diff
@@ -385,3 +385,48 @@
       (if (ctk/in-component-copy? parent)
         true
         (has-any-copy-parent? objects (:parent-id shape))))))
+
+(defn has-any-main?
+  "Check if the shape has any children or parent that is a main component."
+  [objects shape]
+  (let [instance-root (get-instance-root objects shape)]
+    (if (or (ctk/main-instance? shape) (ctk/main-instance? instance-root))
+      true
+      (some true? (map #(has-any-main? objects (get objects %)) (:shapes shape))))))
```

Comment:
I think that this function may do more loops than needed. Perhaps it would be more efficient something like:
1) Get instance root.
2) Get all children with `cfh/get-children-with-self`
3) Iterate all children looking for a main